pkg/virt-handler/multipath-monitor/multipath_monitor.go

```diff
@@ -0,0 +1,197 @@
+package multipath_monitor
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/safepath"
+	"kubevirt.io/kubevirt/pkg/storage/reservation"
+	"kubevirt.io/kubevirt/pkg/virt-handler/filewatcher"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+	virt_chroot "kubevirt.io/kubevirt/pkg/virt-handler/virt-chroot"
+)
+
+const (
+	socket = "multipathd.socket"
+)
+
+//go:generate mockgen -source $GOFILE -package=$GOPACKAGE -destination=generated_mock_$GOFILE
+type mounter interface {
+	Mount(sourcePath, targetPath *safepath.Path, ro bool) *exec.Cmd
+	Umount(path *safepath.Path) *exec.Cmd
+	IsMounted(mountPoint *safepath.Path) (isMounted bool, err error)
+}
+
+type mountManager struct{}
+
+func (m *mountManager) Mount(sourcePath, targetPath *safepath.Path, ro bool) *exec.Cmd {
+	return virt_chroot.MountChroot(sourcePath, targetPath, ro)
+}
+
+func (m *mountManager) Umount(path *safepath.Path) *exec.Cmd {
+	return virt_chroot.UmountChroot(path)
+}
+
+func (m *mountManager) IsMounted(mountPoint *safepath.Path) (isMounted bool, err error) {
+	return isolation.IsMounted(mountPoint)
+}
+
+type MultipathSocketMonitor struct {
+	Watcher             *filewatcher.FileWatcher
+	MultipathSocketPath string
+	HostDir             string
+	Mounter             mounter
+	done                chan struct{}
+}
+
+func NewMultipathSocketMonitor() *MultipathSocketMonitor {
+	const procRootRun = "/proc/1/root/run"
+	multipathSocketPath := filepath.Join(procRootRun, socket)
+	return &MultipathSocketMonitor{
+		Watcher:             filewatcher.New(multipathSocketPath, 1*time.Second),
+		MultipathSocketPath: multipathSocketPath,
+		HostDir:             reservation.GetPrHelperHostSocketDir(),
+		Mounter:             &mountManager{},
+	}
+}
+
+func (m *MultipathSocketMonitor) Run() {
+	m.done = make(chan struct{})
+
+	// Remove old bind mount if there was one
+	if err := m.unmount(); err == nil {
```

Comment:
Could this be a problem if virt handler crashes/restarts and the mount is in use?

Reply:
@mhenriks no, it should clean up the mount at next restart. Additionally, the path is on a hostPath volumes, afaik, kubelet doesn't clean up the directories.

Reply:
I also try to restart the handler and it was restarting without issues

Reply:
Correct me if I'm wrong, but isn't it possible that launchers could be communicating over this socket while it is deleted? That is not a problem?

Reply:
It isn't the launcher but rather the pr-helper which communicates with the multipathd. It could happen that multipathd is restarted, e. g. by systemd, even if I think it should rarely happen when the entire node is restarted. However, if this happens that the scsi reservation in the guest will fail, but once the multipathd is restarted then, the SCSI commands while succeeds. This is something we cannot really control, and IMO not really a concern. Applications which uses scsi reservations, in the guest, needs also to deal with possible temporary failures, and we are in this case.

Reply:
What we could ask is to make the qemu-pr-helper more resilient to failure and to retry a couple of time to connect to multipathd, but this is something which needs to be requested in QEMU code